### ISSUE TITLE
Support for TCP

### ISSUE THREAD

**Issue by jar349** [BuoyantIO/linkerd#806]
I suggest adding support for TCP so that non-HTTP-based services can also leverage linkerd.

My specific use case is for supporting redis, but there are other TCP-based services that would benefit from TCP support.

Whether it makes sense to target redis, or to support arbitrary TCP I leave up to you.

**Comment by iMartyn** [comment]
For consistency it would be nice to only have one ingress to care about in kubernetes, so for your average VPS for instance you'd be looking at smtp(+s), imaps, as well as http and https and auxiliary services.

**Comment by forsandeep** [comment]
I agree with jar349.It will solve many more problems and help running operations even more smoothly

**Comment by Ashald** [comment]
Each time when I think about it there is 1 idea that comes to my mind each time: IIRC, in HTTP proxy spec there is a part that says that for the sake of HTTPS support a client can use a CONNECT method with some special parameters (headers?). In return proxy must figure out the proper destination, open a TCP connection to it and reply with a local endpoint info (IP + port?) that client should use to send arbitrary data as through normal TCP connection. I'm wondering whether this can be used as a "simple" solution to handle TCP traffic. If it's indeed possible I could imagine some kind of micro-utility that would wrap a target process and request required endpoints for it. For instance, let's say we want to talk to Redis that is known as `/db/redis` - then the assumed usage would be something like:
```
$/ request-service --name=/db/redis --output=REDIS_ENDPOINT -- ./my-program-that-uses-redis
```
In this example `request-service` is responsible requesting an "endpoint" from a proxy (Linkerd) and exporting it as env var `REDIS_ENDPOINT` (let's say, it's value is `127.0.0.1:42356`). After this it can just `exec` whatever is given after `--`. In turn, `./my-program-that-uses-redis` is expected to read `REDIS_ENDPOINT` env var and connect to Redis using provided endpoint.

This, definitely, won't allow such fancy stuff as dynamic load balancing and so on but at least will be able to 100% rely on Linkerd for service discovery. In terms of forwarding TCP traffic from local port to remote - that expected to be easy.

I'm wondering how this idea sounds to others? Do you think it's something feasible or rather sounds crazy? The only other "good" (IMHO) approach would be to implement SOCKS5 proxy that (a) allows proxying of arbitrary traffic, (b) can learn about clients original desired destination (e.g., replacement for `host` header on a lower level) and (c) IIRC has some capabilities for protocol negotiation.
